src/structs.py

```diff
@@ -303,3 +309,92 @@ class _Option:
     # A termination condition maps a state to a bool, which is True
     # iff the option should terminate now.
     terminal: Callable[[State], bool] = field(repr=False)
+
+
+@dataclass(frozen=True, repr=False, eq=False)
+class Operator:
+    """Struct defining an operator (as in STRIPS). Lifted!
+    """
+    name: str
+    parameters: Sequence[Variable]
+    preconditions: Collection[LiftedAtom]
+    add_effects: Collection[LiftedAtom]
+    delete_effects: Collection[LiftedAtom]
+    option: ParameterizedOption
+    # A sampler maps a state and objects to a option parameters.
+    _sampler: Callable[[State, Sequence[Object]], ArrayLike] = field(repr=False)
+
+    @cached_property
+    def _str(self) -> str:
+        return f"""{self.name}:
+    Parameters: {self.parameters}
+    Preconditions: {self.preconditions}
+    Add Effects: {self.add_effects}
+    Delete Effects: {self.delete_effects}
+    Option: {self.option}"""
+
+    @cached_property
+    def _hash(self) -> int:
+        return hash(str(self))
+
+    def __str__(self) -> str:
+        return self._str
+
+    def __repr__(self) -> str:
+        return str(self)
+
+    def __hash__(self) -> int:
+        return self._hash
+
+    def __eq__(self, other) -> bool:
+        return str(self) == str(other)
+
+    def ground(self, objects: Sequence[Object]) -> _GroundOperator:
+        """Ground into a _GroundOperator, given objects.
+        """
+        assert len(objects) == len(self.parameters)
+        assert all(o.type == p.type for o, p in zip(objects, self.parameters))
+        sub = dict(zip(self.parameters, objects))
+        preconditions = {atom.ground(sub) for atom in self.preconditions}
+        add_effects = {atom.ground(sub) for atom in self.add_effects}
+        delete_effects = {atom.ground(sub) for atom in self.delete_effects}
+        sampler = lambda s: self._sampler(s, objects)
+        return _GroundOperator(self, objects, preconditions, add_effects,
+                               delete_effects, self.option, sampler)
+
+
+@dataclass(frozen=True, repr=False, eq=False)
+class _GroundOperator:
+    """A ground operator is an operator + objects."""
+    operator: Operator
```

Comment:
Is there a reason GroundOperators need to store a pointer to the lifted Operator? I'm fine with it if we have a reason, but right now it looks like we're just using that lifted Operator's name, so we could just pass the name in instead?

Reply:
I can't think of a reason but I would like to keep it